File: ruby/death_rolls.rb
class Player
  @@number_of_players = 0

  attr_accessor :name, :last_roll
  
  def initialize
    @@number_of_players += 1
    @name = "Player #{@@number_of_players}"
    @last_roll = 1000
  end

  def roll starting_roll
    @last_roll = rand(1..starting_roll).floor
    puts "Random! #{@name} rolls a #{@last_roll} (out of #{starting_roll})."
  end
end

current_player, other_player = Player.new, Player.new

until other_player.last_roll == 1 do
  current_player.roll other_player.last_roll
  current_player, other_player = other_player, current_player
end

puts "#{current_player.name} wins!"

File: go/death_rolls.go
package main

import (
	"fmt"
	"math/rand"
)

var numberOfPlayers = 0

type Player struct {
	name     string
	lastRoll int32
}

func NewPlayer() Player {
	numberOfPlayers++
	return Player{fmt.Sprintf("Player %d", numberOfPlayers), 1000}
}

func (p *Player) Roll(startingNumber int32) {
	p.lastRoll = rand.Int31n(startingNumber) + 1
	fmt.Printf("Random! %s rolls a %d (out of %d).\n", p.name, p.lastRoll, startingNumber)
}

func main() {
	currentPlayer, otherPlayer := NewPlayer(), NewPlayer()

	for otherPlayer.lastRoll != 1 {
		currentPlayer.Roll(otherPlayer.lastRoll)
		currentPlayer, otherPlayer = otherPlayer, currentPlayer
	}

	fmt.Printf("Game over! %s wins.\n", currentPlayer.name)
}
